rs/embedders/tests/system_api.rs

```diff
@@ -2191,3 +2256,237 @@ fn get_system_api_for_best_effort_response(
+    ));
+}
+
+// Helper function to copy test data to heap
+fn copy_to_heap(heap: &mut [u8], data: &[u8]) {
+    heap[..data.len()].copy_from_slice(data);
+}
+
+#[test]
+fn test_env_var_value_operations() {
+    let cycles_account_manager = CyclesAccountManagerBuilder::new().build();
+
+    let var_name_1 = "TEST_VAR_1".to_string();
+    let var_name_path = "PATH".to_string();
+    let var_name_empty = "EMPTY_VAR".to_string();
+    let var_value_1 = "Hello World".to_string();
+    let var_value_path = "/usr/local/bin:/usr/bin".to_string();
+
+    let mut env_vars = BTreeMap::new();
+    env_vars.insert(var_name_1.clone(), var_value_1.clone());
+    env_vars.insert(var_name_empty.clone(), "".to_string());
+    env_vars.insert(var_name_path.clone(), var_value_path.clone());
+
+    let api = get_system_api(
+        ApiTypeBuilder::build_update_api(),
+        &SystemStateBuilder::default()
+            .environment_variables(env_vars)
+            .build(),
+        cycles_account_manager,
+    );
+    let mut heap = vec![0u8; 256];
+
+    // Test ic0_env_var_count.
+    assert_eq!(api.ic0_env_var_count().unwrap(), 3);
+
+    // Test copying value for null bytes as variable name.
+    assert_eq!(
+        api.ic0_env_var_value_copy(0, var_name_empty.len(), 0, 0, 0, &mut heap),
+        Err(HypervisorError::EnvironmentVariableNotFound {
+            name: String::from_utf8(heap[0..var_name_empty.len()].to_vec()).unwrap()
+        })
+    );
+
+    // Test empty variable.
+    copy_to_heap(&mut heap, var_name_empty.as_bytes());
+    assert_eq!(
+        api.ic0_env_var_value_size(0, var_name_empty.len(), &heap)
+            .unwrap(),
+        0
+    );
+
+    // Test copying an empty variable.
+    api.ic0_env_var_value_copy(0, var_name_empty.len(), 0, 0, 0, &mut heap)
+        .unwrap();
+    let mut expected_empty_value = vec![0u8; 64];
+    copy_to_heap(&mut expected_empty_value, var_name_empty.as_bytes());
+    assert_eq!(&heap[0..64], expected_empty_value);
+
+    // Test value size and copying for existing variable: TEST_VAR_1.
+    copy_to_heap(&mut heap, var_name_1.as_bytes());
+    assert_eq!(
+        api.ic0_env_var_value_size(0, var_name_1.len(), &heap)
+            .unwrap(),
+        var_value_1.len() // length of "Hello World"
+    );
+
+    api.ic0_env_var_value_copy(0, var_name_1.len(), 0, 0, var_value_1.len(), &mut heap)
+        .unwrap();
+    assert_eq!(&heap[0..11], var_value_1.as_bytes());
+
+    // Test value size and copying for existing variable: PATH.
+    copy_to_heap(&mut heap, var_name_path.as_bytes());
+    assert_eq!(
+        api.ic0_env_var_value_size(0, var_name_path.len(), &heap)
+            .unwrap(),
+        var_value_path.len() // length of "/usr/local/bin:/usr/bin"
+    );
+    api.ic0_env_var_value_copy(
+        0,
+        var_name_path.len(),
+        0,
+        0,
+        var_value_path.len(),
+        &mut heap,
+    )
+    .unwrap();
+    assert_eq!(&heap[0..23], var_value_path.as_bytes());
```

Comment:
To reduce code duplication, I'd refactor the three tests into a for loop over the three env variables with a uniform body handling the checks for the individual env vars.

Reply:
[Ticket](https://dfinity.atlassian.net/browse/EXC-2070)